SitesDropdown: Use Site IDs internally, drop SitesList from consumers

ockham opened the pull request:
Drop `sites-list` from:
* `components/site-selector-modal`
* `me/account/main`
* `me/help/help-contact-form/index`

Make `SitesDropdown` use siteIds instead of slugs for all relevant props:
* `selectedSiteId` (same as before)
* `onSiteSelect` callback arg
* `filter` callback arg

Consumers are updated accordingly. Also, change internal state representation to IDs instead of mixing IDs and site objects.

Note that `SitesDropdown` was only using site related state/selectors to translate between site IDs, slugs, and objects. This is better overcome at the level of the components that it uses to display site information; thus, this PR changes the `Site` block component to also accept site IDs (instead of just objects). /cc @mcsf 

To test: Test the following instances of `SitesDropdown`:
* `/themes`: The `ThemesSiteSelectorModal`. Click on any theme's '...' menu, and select an action item (like 'Activate'). Select a site, and proceed. The action should be executed for that selected site.
* `/me/account`: Verify that changing your primary website works.
* `/help/contact`: Select a site you need help with, spam HEs 😛

Follow-up PR:
* SiteSelector
* SiteSelector consumers
  * components/sites-popover
  * my-sites/picker: No `onSiteSelect` or `filter` callback props
  * my-sites/sites/sites

Supersedes #9624. Take two for #13278 (more advanced!)

ockham commented:
As discussed in Slack, we can merge this without further ado once we have a `<QuerySites allSites />`, as provided by #13094.

mtias commented:
@ockham can we resurrect?

ockham commented:
> @ockham can we resurrect?

Yep, we totall should, this time killing all `sites-list` remainders. I'll try to get to it this week. I'll see if it's worth rebasing or if starting fresh will be easier.

ockham commented:
Okay, I'd say this is ready for review.

ockham commented:
BTW the error is eslint complaining about the `Site` block not being an ES6 class. Since this PR is already quite big, I'm inclined to leave as-is.

gwwar commented on client/blocks/site/index.jsx:
```diff
@@ -145,3 +147,10 @@ export default React.createClass( {
 } );
+
+export default connect( ( state, { site } ) => {
+	if ( isNumber( site ) ) { // Accept site IDs
```
This feels a bit strange to me, I think adding `siteId` would be clearer here.

So both 
```jsx
<Site site={ site } />
```

and 
```jsx
<Site siteId={ siteId } />
```

would be accepted usages. We can then do:
```jsx
const mapStateToProps = ( state, { siteId, site } ) => {
    const site = siteId ? getSite( state, site ) : site;
    return {
        site
    };
}
```

ockham replied:
Can do. The rationale for using `site` both for objects and IDs was to guard against unpredictable/unintuitive behavior when providing both a `site` and a `siteId` prop. But that's probably not really a biggie, so I'll go ahead with what you suggested.

gwwar commented on client/components/site-selector-modal/index.jsx:
```diff
@@ -35,45 +34,46 @@ class SiteSelectorModal extends Component {
-		const primarySite = sitesList.getPrimary();
-		let filteredSites = sitesList.getVisible();
+		const { primarySiteId, visibleSites } = this.props;
+		let filteredSiteIds = visibleSites.map( ( site ) => site.ID );
```
This feels selectorish. Since this is done in the constructor, what happens if visibleSites changes?

ockham replied:
Ah, good point. I should probably move it to `mapStateToProps`.

gwwar commented on client/blocks/site/index.jsx:
```diff
@@ -145,3 +147,10 @@ export default React.createClass( {
+	if ( isNumber( site ) ) { // Accept site IDs
+		return { site: getSite( state, site ) };
+	}
+	return { site };
```
Note that since we're also overriding site, we'll only be fetching the site once, which may leave a stale site in this component.

gwwar commented:
Overall this appears to work well. Let me know what you think about not overloading the site property.

@hoverduck would it be possible to run the e2e tests against this branch?

hoverduck commented:
@gwwar e2e tests all green

gwwar commented:
Thanks @hoverduck !

ockham commented:
Thanks a lot for reviewing @gwwar (and @hoverduck for running e2e tests)! I've rebased and updated the branch according to your suggestions. Okay to 🚢 ?

tyxla commented on client/blocks/site/index.jsx:
```diff
@@ -11,8 +12,9 @@ import Gridicon from 'gridicons';
 
-export default React.createClass( {
+const Site = React.createClass( {
 	displayName: 'Site',
```
We won't need the `displayName` now, right?

ockham replied:
Ah, good catch. I'll remove it.

tyxla commented on client/state/selectors/test/get-selected-or-primary-site-id.js:
```diff
@@ -0,0 +1,39 @@
+	} );
+
+	context( 'with a site selected', () => {
+		it( 'should return the selected site\'s ID', ( ) => {
```
Does it make sense to also add a case with selected site, current user and primary site, just so we actually test the priority of `getSelectedOrPrimarySiteId`?

ockham replied:
Yeah, I was considering that too, but I wasn't sure how to group tests for that scenario. Maybe I'll just add a current user and primary site to this test case... :thinking:

ockham replied:
Ah nvm, I'll just structure tests differently...

tyxla commented on client/blocks/site/README.md:
```diff
@@ -9,13 +9,14 @@ This component displays a Site item using the data structure provided by sites-l
 
-* `site (object)` - A site object from `sites-list`. It's required.
+* `siteId (number)` - A site ID. Required, unless a `site` object prop is provided (see below).
+* `site (object)` - A site object. Required, unless a `siteId` prop is provided.
```
Looking at the docs, it's not clear whether `siteId` or `site` would be treated with higher priority. Should we indicate that somewhere? I know it wouldn't make sense to pass both props at the same time, but just in case 😉

ockham replied:
Good point (see my [comment](https://github.com/Automattic/wp-calypso/pull/13293#discussion_r121366048) :slightly_smiling_face: ) -- I'll add a note.

tyxla commented on client/components/site-selector-modal/index.jsx:
```diff
@@ -98,11 +85,28 @@ class SiteSelectorModal extends Component {
+		return {
+			initialSiteId: includes( filteredSiteIds, primarySiteId )
+				? primarySiteId
+				: filteredSiteIds[ 0 ]
```
Any chance to have no visible sites at the moment?

ockham replied:
I... hope not? 😄 We've had this part of the logic [before](https://github.com/Automattic/wp-calypso/pull/13293/files#diff-f7f3055cc9ffe5ea8a7a5a5712368a8eL54), so I'm not too worried, TBH.
